book/layout.md

```diff
@@ -25,117 +25,124 @@ width and height, is never computed. That makes it pretty hard to draw
-============
-
-Different layout objects lay out their children differently. Some
-*container* elements, like paragraphs, contain text, and lay that text
-out horizontally in lines. But other elements contain such containers
-themselves, which are stacked vertically. In the example above, the
-body element contains a heading and some paragraphs, which are
-themselves containers for text.
-
-Abstracting a bit, there are two *layout modes* here, two ways an
-element can be laid out relative to its children.[^or-equivalently]
-They're called block layout, for laying out containers, and inline
-layout, for laying out text. In the example above the body element is
-in block layout mode (because its children are containers) while the
-paragraphs and heading are in inline layout mode (because their
-children are text). In real browsers there are a lot of other layout
-modes too, but in this chapter we'll keep it simple.
-
-[^or-equivalently]: The oldest CSS properties that affect layout mode,
-like the `inline` and `block` values for `display`, are set on
-children, which is confusing. Newer ones like `inline-block`, `flex`,
-and `grid` are set on the parent. This chapter uses the newer
-convention, even though it's actually implementing inline and block
-layout.
-
-Our browser will have one class for each layout mode. `InlineLayout`
-should be familiar: it's what we've been implementing up until this
-point and corresponds to how text inside a paragraph are laid out in
-lines, left to right.[^in-english] `BlockLayout` is new to this
-chapter: it corresponds to the way containers like paragraphs and
-headings are laid out one after another vertically from top to bottom.
+``` {.python}
+class BlockLayout:
+    def __init__(self, node, parent, previous):
+        self.node = node
+        self.parent = parent
+        self.previous = previous
+        self.children = []
+```
 
-[^in-english]: In European languages, at least!
+Each layout object also needs a size and position, which we'll store
+in `width`, `height`, `x`, and `y` fields. But let's leave that for
+later. The first job for `BlockLayout` is creating the layout tree
+itself.
 
-Since the existing `Layout` class is more or less inline layout, let's
-rename it to `InlineLayout` and rename its constructor to be a new
-`layout` method:
+We'll do that in a new `layout` method, looping over each child _node_
+and creating a new child _layout object_ for it.
 
 ``` {.python}
-class InlineLayout:
+class BlockLayout:
     def layout(self):
-        # ...
+        previous = None
+        for child in self.node.children:
+            next = BlockLayout(child, self, previous)
+            self.children.append(next)
+            previous = next
 ```
 
-So this renamed `layout` method initializes `weight`, `style`, and
-`size` fields, as well as the `display_list`:
+The difference between _nodes_ and _layout objects_ can get confusing
+here. Each `child` comes from `node.children`, so it is a HTML node.
```

Comment:
Your comments here might be creating as much confusion as they aim to solve, e.g. because self.children actually is layout objects whereas self.node.children are nodes. I was confused myself for about 30 seconds. How about instead putting\ some comments in the code? Something like:

def layout(self); 
  previous = None
  for child in self.node.children:
    // Note: child is a node, not a layout object! Likewise self.node.children is an array of nodes
    next = BlockLayout(...)
    self.children.append(next)
    previous = next

Reply:
WDYT of my comments?